Feature/ExtendedHubbardMom1D

Skuwar1 opened the pull request:
# `ExtendedHubbardMom1D`
- A new model (i.e. ``ExtendedHubbardMom1D``) is added to the Rimu.
- The model is an extended version of ``HubbardMom1D`` in Rimu.
- Furthermore, the model is equipped with a ```theta```-twisted boundary condition inspired by the "ExtendedHubbardReal1D".

Skuwar1 committed:
Create ExtendedHubbardMom1D.jl

- New model (i.e. ``ExtendedHubbardMom1D`` is added to Rimu

Skuwar1 committed:
Update excitations.jl

- ``extended_momentum_transfer_diagonal`` is added for the nearest neighbour term in ``ExtendedHubbardMom1D``

Skuwar1 committed:
Update Hamiltonians.jl

- ``ExtendedHubbardMom1D`` is added to all the necessary test sets.

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update hamiltonians.md

Skuwar1 committed:
Update Hamiltonians.jl

joachimbrand commented on src/Hamiltonians/excitations.jl:
```diff
@@ -79,10 +79,6 @@ See [`excitation`](@ref), [`OccupiedModeMap`](@ref).
     return excitation(add, dst_indices, src_indices)..., src_modes..., -mom_change
 end
 
-function momentum_transfer_excitation(add::FermiFS, chosen::Integer, map; fold=true)
-    return add, 0.0, 0, 0, 0
-end
-
```
Why was this function removed?

Skuwar1 replied:
I'm sorry. I didn't notice that this function was necessary for other operations. I have added it back. I have also defined ``extended_momemtum_transfer_excitation``as a copy of ``momentum_transfer_excitation(:: BoseFS)``.

joachimbrand commented on src/Hamiltonians/excitations.jl:
```diff
@@ -139,6 +135,27 @@ function momentum_transfer_diagonal(map::BoseOccupiedModeMap)
     return float(onproduct)
 end
+
+"""
+    extended_momentum_transfer_diagonal(map, M)
+
+The diagonal part of nearest neighbour [`momentum_transfer_excitation`](@ref).
+"""
+
```
Please remove the empty line 144! Otherwise the docstring won't be found.

It would also be nice to have a bit more verbose explanation of what the arguments mean and what is returned.

joachimbrand commented on src/Hamiltonians/excitations.jl:
```diff
@@ -139,6 +135,27 @@ function momentum_transfer_diagonal(map::BoseOccupiedModeMap)
+The diagonal part of nearest neighbour [`momentum_transfer_excitation`](@ref).
+"""
+
+function extended_momentum_transfer_diagonal(map::OccupiedModeMap,M::Int)
+    onproduct = 0
+    step = (2*π)/M
```
```suggestion
function extended_momentum_transfer_diagonal(map::OccupiedModeMap, step)
    onproduct = 0
```
I would suggest to pre-calculate `step` in the calling function, where it can be done by the compiler. In this function the number of modes is run-time data and the (expensive) floating-point division would have to be done at run-time.

joachimbrand commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,128 @@
+    h::ExtendedHubbardMom1D{<:Any,M,<:BoseFS}, map
+) where {M}
+    return h.u / 2M *momentum_transfer_diagonal(map) 
+    + h.v/ M * dot(cos.(h.ks), map) * extended_momentum_transfer_diagonal(map,M) 
```
```suggestion
    + h.v/ M * dot(cos.(h.ks), map) * extended_momentum_transfer_diagonal(map, (2*π)/M) 
```
See comment on `extended_momentum_transfer_diagonal`!

joachimbrand commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,128 @@
+"""
+    ExtendedHubbardMom1D(address; u=1.0, t=1.0, v=1.0, dispersion=hubbard_dispersion, boundary_condition = 0.0)
+
+Implements a one-dimensional extended Hubbard chain in momentum space.
```
```suggestion
Implements a one-dimensional extended Hubbard chain, also known as the ``t - V`` model, 
in momentum space.
```

joachimbrand commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,128 @@
+"""
+    ExtendedHubbardMom1D(address; u=1.0, t=1.0, v=1.0, dispersion=hubbard_dispersion, boundary_condition = 0.0)
```
```suggestion
    ExtendedHubbardMom1D(
        address; 
        u=1.0, t=1.0, v=1.0, dispersion=hubbard_dispersion, boundary_condition = 0.0
    )
```
Please break the long lines over multiple lines. The recommendation is to not exceed 72 characters per line to help readability. You can set vs-code to display a ruler in order to help you see quickly where the 72 character margin lies. See https://github.com/JuliaDiff/BlueStyle?tab=readme-ov-file#vs-code-settings

joachimbrand commented on test/Hamiltonians.jl:
```diff
@@ -173,6 +173,9 @@ end
         ExtendedHubbardReal1D(BoseFS((1, 0, 0, 0, 1)); u=1.0, v=2.0, t=3.0),
         ExtendedHubbardReal1D(BoseFS(1, 0, 2, 1); u=1 + 0.5im),
         ExtendedHubbardReal1D(BoseFS(1, 0, 2, 1); t=1 + 0.5im),
+        ExtendedHubbardMom1D(BoseFS((1, 0, 0, 0, 1)); u=1.0, v=2.0, t=3.0),
+        ExtendedHubbardMom1D(BoseFS(1, 0, 2, 1); u=1 + 0.5im),
+        ExtendedHubbardMom1D(BoseFS(1, 0, 2, 1); t=1 + 0.5im),
```
It would be good to add testing with `FermiFS` and `OccupationNumberFS` addresses as well.

joachimbrand commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,128 @@
+    return address, -((ham.v * cos((q)*((2*π)/M)))/(M))*onproduct
+end
+
+function offdiagonals(h::ExtendedHubbardMom1D, a::SingleComponentFockAddress)
+    map = OccupiedModeMap(a)
+    singlies = length(map)
+    doublies = count(i -> i.occnum ≥ 2, map)
+    num = num_offdiagonals(h, a, singlies, doublies)
+    return OffdiagonalsBoseMom1D(h, a, num, map)
+end
```
Hmm, this looks a bit like unnecessary duplication as it is identical to the code for `HubbardMom1D` and could be eliminated by changing the dispatch rules for the latter. Maybe it is still valuable (for readability) to have the code here?

We should also think about whether it makes sense to only have `HubbardMom1D`, where the extended version is chosen if `v != 0`.

joachimbrand reviewed (changes requested):
Looks very good already.

I've mostly got a few pedantic niggles.

Can we make this work with `OccupationNumberFS`? It's not a top priority, but it also shouldn't be hard to do, as `HubbardMom1D` already works with this address type.

```julia-repl
julia> oaddr = OccupationNumberFS(1,2,0,0)
OccupationNumberFS{4, UInt8}(1, 2, 0, 0)

julia> sparse(HubbardMom1D(oaddr))
5×5 SparseArrays.SparseMatrixCSC{Float64, Int64} with 21 stored entries:
 -2.75      0.707107  0.5       0.25       ⋅ 
  0.707107  1.5       0.353553  0.707107  0.612372
  0.5       0.353553  1.25      0.5       0.433013
  0.25      0.707107  0.5       5.25       ⋅ 
   ⋅        0.612372  0.433013   ⋅        0.75

julia> sparse(ExtendedHubbardMom1D(oaddr))
ERROR: MethodError: no method matching momentum_transfer_diagonal(::ExtendedHubbardMom1D{…}, ::OccupiedModeMap{…})
```

Skuwar1 committed:
Update src/Hamiltonians/ExtendedHubbardMom1D.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ExtendedHubbardMom1D.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/excitations.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ExtendedHubbardMom1D.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update excitations.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update excitations.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

joachimbrand committed:
docstring fix for HubbardMom1DEP

Skuwar1 committed:
Update HubbardMom1DEP.jl

mtsch commented on src/Hamiltonians/excitations.jl:
```diff
@@ -83,6 +83,60 @@ function momentum_transfer_excitation(add::FermiFS, chosen::Integer, map; fold=t
     return add, 0.0, 0, 0, 0
 end
 
+function extended_momentum_transfer_excitation(add::SingleComponentFockAddress, chosen::Integer, map; fold=true)
```
This is the same function as `momentum_transfer_excitation`. Delete this one and use `momentum_transfer_excitation` instead.

Skuwar1 replied:
That's true, but the problem is that the function ``momentum_transfer_excitation(address:: FermiFS)``is also defined, which returns ``new_addr, 0.0, 0.0, 0.0, 0.0``, which we do not want. That is the reason why I defined the same thing with a different name.

Skuwar1 replied:
Please let me know if there is a more efficient way.

mtsch commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,124 @@
+    return singlies * (singlies - 1) * (M - 2) + doublies * (M - 1)
+end
+
+@inline function momentum_transfer_diagonal(
+    h::ExtendedHubbardMom1D{<:Any,M,<:SingleComponentFockAddress}, map
+) where {M}
+    return (h.u/ 2M) * momentum_transfer_diagonal(map) 
+    + (h.v/ M) * extended_momentum_transfer_diagonal(map, (2*π)/M) 
+end
+
+@inline function momentum_transfer_diagonal(
+    h::ExtendedHubbardMom1D{<:Any,M,<:FermiFS}, map
+) where {M}
+    return (h.v/ M) * extended_momentum_transfer_diagonal(map, M)
+end
```
This might be how it's done elsewhere, but I'd prefer if you overload `diagonal_element` instead.

Skuwar1 committed:
Update excitations.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

Skuwar1 committed:
Update HubbardMom1D.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

mtsch commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,117 @@
+    ham::ExtendedHubbardMom1D{<:Any,M,A}, address::A, chosen, map=OccupiedModeMap(address)
+) where {M,A<:SingleComponentFockAddress}
+    address, onproduct,_,_,q = momentum_transfer_excitation(address, chosen, map)
+    return address, ham.u/(2*M)*onproduct + ((ham.v * cos((q)*((2*π)/M)))/(M))*onproduct
```
```suggestion
    return address, ham.u * onproduct / 2M + ham.v * cos(q * 2π / M) * onproduct / M
```
Clean up expressions like these so they're easier to read

mtsch commented on src/Hamiltonians/ExtendedHubbardMom1D.jl:
```diff
@@ -0,0 +1,117 @@
+    ham::ExtendedHubbardMom1D{<:Any,M,A}, address::A, chosen, map=OccupiedModeMap(address)
+) where {M,A<:FermiFS}
+    address, onproduct,_,_,q = momentum_transfer_excitation(address, chosen, map)
+    return address, -((ham.v * cos((q)*((2*π)/M)))/(M))*onproduct
```
```suggestion
    return address, -ham.v * onproduct * cos(q * 2π / M) / M
```

mtsch reviewed (changes requested):
Some formatting suggestions. Apart from that and some more tests, it looks good!

Skuwar1 committed:
Update Hamiltonians.jl

- Added test for comparison between energies of ``ExtendedHubbardMom1D`` and ``ExtendedHubbardReal1D``

Skuwar1 committed:
Update src/Hamiltonians/ExtendedHubbardMom1D.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

Skuwar1 committed:
Update src/Hamiltonians/ExtendedHubbardMom1D.jl

Co-authored-by: mtsch <matijacufar@gmail.com>

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

mtsch commented on test/Hamiltonians.jl:
```diff
@@ -1880,3 +1880,12 @@ end
+    addr = FermiFS{3,6}(0,1,1,1,0,0)
+    for boundary_condition in ([i*π for i in 0.0:0.2:1.0])
+        HR = ExtendedHubbardMom1D(addr; boundary_condition)
+        HM = ExtendedHubbardReal1D(addr; boundary_conition= boundary_condition/6)
```
```suggestion
        HM = ExtendedHubbardMom1D(addr; boundary_condition)
        HR = ExtendedHubbardReal1D(addr; boundary_conition=boundary_condition/6)
```
Why does the real space version need `boundary_condition / 6`? I think it should be the same for both Hamiltonians. It would also be good to repeat the test with `BoseFS`.

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

joachimbrand commented on src/Hamiltonians/excitations.jl:
```diff
@@ -139,6 +135,26 @@ function momentum_transfer_diagonal(map::BoseOccupiedModeMap)
+    extended_momentum_transfer_diagonal(map, step)
+
+The diagonal part of nearest neighbour term [`momentum_transfer_excitation`](@ref) in [`ExtendedHubbardMom1D`](@ref).
+Where `step` is the separation between the two nearest single-particle momentum.
```
```suggestion
Where `step` is the separation of single-particle momenta in the momentum grid.
```

joachimbrand reviewed (changes requested):
Only a minor rewording suggestion for a docstring. Otherwise this looks good 😄

Skuwar1 committed:
Update src/Hamiltonians/excitations.jl

Co-authored-by: Joachim Brand <joachim.brand@gmail.com>

joachimbrand reviewed (dismissed):
Thanks! It looks good now.

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update Hamiltonians.jl

Skuwar1 committed:
Update excitations.jl

Skuwar1 committed:
Update ExtendedHubbardMom1D.jl

Skuwar1 committed:
Update excitations.jl

Skuwar1 committed:
Update excitations.jl

mtsch reviewed (approved):
Looks good to me!
